src/resolvers/Mutation/assignTag.ts

```diff
@@ -0,0 +1,76 @@
+import { MutationResolvers } from "../../types/generatedGraphQLTypes";
+import { errors, requestContext } from "../../libraries";
+import { User, Tag, UserTag } from "../../models";
+import {
+  USER_NOT_FOUND_ERROR,
+  USER_NOT_AUTHORIZED_ERROR,
+  TAG_NOT_FOUND,
+} from "../../constants";
+
+export const assignTag: MutationResolvers["assignTag"] = async (
+  _parent,
+  args,
+  context
+) => {
+  const currentUser = await User.findOne({
+    _id: context.userId,
+  }).lean();
+
+  const requestUser = await User.exists({
+    _id: args.userId,
+  });
+
+  // Checks whether currentUser and the requestUser both exist.
+  if (!currentUser || !requestUser) {
+    throw new errors.NotFoundError(
+      requestContext.translate(USER_NOT_FOUND_ERROR.MESSAGE),
+      USER_NOT_FOUND_ERROR.CODE,
+      USER_NOT_FOUND_ERROR.PARAM
+    );
+  }
+
+  // Get the tag object
+  const tag = await Tag.findOne({
+    _id: args.tagId,
+  }).lean();
+
+  if (!tag) {
+    throw new errors.NotFoundError(
+      TAG_NOT_FOUND.MESSAGE,
+      TAG_NOT_FOUND.CODE,
+      TAG_NOT_FOUND.PARAM
+    );
+  }
+  // Check that the user shouldn't already be assigned the tag
+  const userAlreadyHasTag = await UserTag.exists({
```

Comment:
@EshaanAgg @xoldyckk I have a doubt here. As we are only dealing with the tags (no tagFolders anymore), so, what if we pass the child TagId here, we'll assign the child tagId here, even if he's already assigned the parent tagId.
To resolve this issue, we have 2 approaches that we can follow here
- Either we assign user all the child Tags too, when any parent Tag is assigned.
- Either we add a check here, to check if the tagId belongs to any child TagId of the previously assigned parent tagId.
What should be done?

Reply:
I think this was implemented for the previously thought approach of `tagFolder`.

Reply:
@beingnoble03 I haven't reviewed anything except mongoose models and graphql typeDefs. Let's first have a proper strucuture to work with.

Regarding your question I'm not sure myself. It'll be left as a decision for @palisadoes.

In gmail, if you assign something a parent label, the parent label along with all its child labels, no matter how deeply nested are also applied to the email. I explained logic for this over [here](https://github.com/PalisadoesFoundation/talawa-api/issues/1126#issuecomment-1462372110).

If @palisadoes wants we can put together a logic to only apply labels if they have no successive child. The logic will work like this:-

```
school

school/class1/mathematics
school/class1/science

school/class2
```

So, if you assign someone a label of `school` or `school/class1`, it'll fail. It'll only work for `school/class1/mathematics` and `school/class1/science`. It'll also work for `school/class2` as it has no successive child.

Reply:
It makes more sense to follow the Gmail like logic in this case. 

As from the implementation point of view, any user would just be assigned their immediate parent in the database. Then when queries with a tag, we can use graph lookup to first query all of the child tags, and then make another query to fetch all the users belonging to any of those parent or child tags.

A similar logic could be implemented for users as well, but I feel from the frontend perspective, it would make the most sense to return the deepest tags only (as we might to show 'married' tag on the user's profile, but showing the 'marital status' tag makes no sense). However the same could be queried with the help of the current fields if the need arises.

Reply:
@EshaanAgg 

If you're talking about how tags will work in talawa-admin for filtering and stuff, they'll work how the image for breeze cms shows us. First you fetch list of parent tags(it'll be paginated using graphql connection), when you click on the tag you see all users who are related to that tag. There'll be an **_plus(+)_** button on the tag, when you click it, it'll load its child tags. Clicking on each child tag will reveal users related to that tag. This approach will be followed for however deeply nested the tags are, until you reach the last nested tag level. Like this:-

_**school +**_

_**school -**_
&nbsp;&nbsp; _**class1 +**_
&nbsp;&nbsp;**_class2_ +**

_**school -**_
&nbsp;&nbsp; _**class1 -**_
&nbsp;&nbsp;&nbsp;&nbsp;_**science**_
&nbsp;&nbsp;&nbsp;&nbsp; _**mathematics**_
&nbsp;&nbsp;**_class2_ -**
&nbsp;&nbsp;&nbsp;&nbsp; _**science**_
&nbsp;&nbsp;&nbsp;&nbsp; **_mathematics_**

Each subsequent nested level will only be loaded when you click on that **_+_** icon. When you click on **_-_** it'll be minimized.

Regarding what you said about only sending the deepest tags, I take it you meant how to display it on a user's profile/information. I don't think that's good. For example you're viewing a user's profile, for a particular organization. The child tag by itself will make no sense.

What you're suggesting:-

**_science_**

This is what I think it should be:-

**_school/class1/science_**

Reply:
Yes @xoldyckk. For the display on the navigation panel, I was thinking the same. The description I had provided earlier was more related to how to how the assigning of roles would work. Like if an event I want to add `class2` as attendees, then to implement the same our resolver would make the two database queries, one to get the complete subtree of the the parent tag (that is `class2`, `class2/science`, `class2/mathematics`), and then other query to fetch the details of all the users assigned to any of these tags.

As for the display, the primary reason I was advocating for the earlier approach  was that as a user starts to be allotted multiple tags, the above way of display gets too cluttered too fast. Anyhow we can decide on the same later, once the definitions and structures are finalised.

Reply:
@EshaanAgg At anytime, the tag's id will be available wherever the tag is used. You only need the tagId to query the users related to it.

I didn't understand what you said, explain with some diagrams man, like I'm making these badass tables.

Reply:
Ahhh I wish I was as badass as you, so I'll just copy yours!

For the given structure:
```
school
   class1 
     science
     mathematics
  class2 
     science
     mathematics
```

Let's say we are working with events (PTM for example), and we want to assign all the `class2` students to be attendees of the same. So in such a case, with the current implementation, the frontend would make to make a singular query of the sort `allAttendee(tag: class2, event: PTM)`. 

On the backend, we would receive the given query and then in effect make two database queries. The first one would be a `graphLookup` query of the `tags` model, that would return the subtree of the given input tag.
In our case, it would be the tags `school/class2`, `school/class2/science` and `school/class2/mathematics`. 

Then we would make a second query, this time to the `TagUser` model, to delete all the entries if their `tag` belongs to the list of the tags generated earlier. A similar approach will also be used in the `removeTag` mutation.

Reply:
@EshaanAgg

First of all frontend needs to make a query with tagId not tagName. What they see on their screen will be tagName, but what gets sent in mutation will be tagId. TagId can be uniquely identified in `TagUser` junction, because we'll have the tagId.

IDK what you mean when you say we need to fetch subtree of tags. We can get all users related to a tag whether it has nested tags or not in a single query. Look at [this](https://github.com/PalisadoesFoundation/talawa-api/issues/1126#issuecomment-1462372110) again. When I say they'll get assigned to parent tags, I mean in `TagUser` table the user will get a new row(document) connected to each of those tags. If a tag has `n` levels of nesting with 2 items for each nesting level, the number of `TagUser` rows(documents)  created will be `2^n - 1`.

If you want to add all users with tag `school/class1` to the event `PTM`, just make the mutation with `school/class1` tag's id.

This is how gmail's lablels works, I know this is too much duplication of data and it can be avoided, but that'll require another complex implementation. This also will perform worse during mutations, the more levels of nesting you have, that's why I asked whether the levels of nesting should be finite. But @palisadoes said it has to be infinite so, I guess not.

Reply:
@EshaanAgg I just tried gmail myself and what I read on google about gmail's labels was wrong. Gmail doesn't automatically assign child labels of a parent label to the email which currently has the parent label assigned to it. So, [this](https://github.com/PalisadoesFoundation/talawa-api/issues/1126#issuecomment-1462372110) explanation of mine is incorrect.

Reply:
@xoldyckk 
1. I know that and would use the `_id` properties of the tag only to work with them (on frontend) as obviously the name's of the tags would not be unique across the different folders, and would be bad practice in general. The choice of using names in the examples was just done for the ease of implementation.

2. Ahh got it. I was advocating for the other approach, where on assigning the tag `school/class2/science` to `userA` would only create a single row in the `TagUser` schema, corresponding to 
```
userA    school/class2/science
```
and not 
```
userA    school
userA    school/class2
userA    school/class2/science
```

The feel the later approach is bad, not because its duplication of data, but it is exponential duplication of data. The whole point of this flexibility is to allow users to nest the tags as deep as they want, and when we keep this consideration in mind, the amount of overhead generated is concerning.

With my approach, although the resolver logic would be difficult to implement, the same would achieve the desired effect with less overall overhead. For example, if we query to get all the users with the tag `class2`, we will first traverse the `subtree` of tags of `class2`, that is all the tags that have this tag as their ancestor (immediate or other). In our case they are `school/class2/science` and `school/class2/mathematics`.
Not we can perform the required operation in the `UserTag` scheme, but instead of using an equality in id, i.e. 
```ts
UserTag.find({
   _id : args.id,
})
```

we will first generate this list of child tags, and then use the `$in` clause:
```ts
const allChildTags = []     // contains school/class2, school/class2/science and school/class2/mathematics 
UserTag.find({
   _id : {
     $in: allChildTags,
    }
})
```

3. Oops. What do you think we should continue with now? Should we follow the gmail based approach, or continue with this `child also gets parent` tag approach.

Reply:
@EshaanAgg Gmail's approach is better. We were scratching our heads uselessly because we thought gmail used some complex checks in their backend or something to make that previous approach work. But their implementation is very simple. Read the link I posted above. What we were doing was premature optimization.

Reply:
Got it. It will now be a simple one-to-one relationship, which would put the burden on client, and not on us.